src/subscriber.ts

```diff
@@ -14,497 +14,351 @@
+    // 1 message at a time.
+    if (options.flowControl) {
+      const {maxMessages = 100} = options.flowControl;
+
+      if (!options.streamingOptions) {
+        options.streamingOptions = {} as MessageStreamOptions;
+      }
+
+      const {maxStreams = 5} = options.streamingOptions;
+      options.streamingOptions.maxStreams = Math.min(maxStreams, maxMessages);
     }
-    return this.flushTimeoutHandle_;
   }
-  /*!
-   * Sets a timeout to modify the ack deadlines for any unacked/unnacked
-   * messages, renewing their lease.
+  /**
+   * Callback to be invoked when a new message is available.
+   *
+   * New messages will be added to the subscribers inventory, which in turn will
+   * automatically extend the messages ack deadline until either:
+   *   a. the user acks/nacks it
+   *   b. the maxExtension option is hit
+   *
+   * If the message puts us at/over capacity, then we'll pause our message
+   * stream until we've freed up some inventory space.
+   *
+   * New messages must immediately issue a ModifyAckDeadline request
+   * (aka receipt) to confirm with the backend that we did infact receive the
+   * message and its ok to start ticking down on the deadline.
    *
    * @private
    */
-  setLeaseTimeout_() {
-    if (this.leaseTimeoutHandle_ || !this.isOpen) {
-      return;
+  private _ondata(response: PullResponse): void {
+    response.receivedMessages.forEach((data: ReceivedMessage) => {
+      const message = new Message(this, data);
+
+      this._inventory.add(message);
+      this.modAck(message, this.ackDeadline);
+    });
+
+    if (this._inventory.isFull()) {
+      this._stream.pause();
+      this._inventory.onFree().then(() => this._stream.resume());
     }
-    const latency = this.latency_.percentile(99);
-    const timeout = Math.random() * this.ackDeadline * 0.9 - latency;
-    this.leaseTimeoutHandle_ =
-        setTimeout(this.renewLeases_.bind(this), timeout);
   }
+
   /**
-   * Writes to specified duplex stream. This is useful for capturing write
-   * latencies that can later be used to adjust the auto lease timeout.
+   * Returns a promise that will resolve once all pending requests have settled.
    *
    * @private
    *
-   * @param {string} connId The ID of the connection to write to.
-   * @param {object} data The data to be written to the stream.
    * @returns {Promise}
    */
-  writeTo_(connId, data) {
-    const startTime = Date.now();
-    return new Promise((resolve, reject) => {
-      this.connectionPool!.acquire(connId, (err, connection) => {
-        if (err) {
-          reject(err);
-          return;
-        }
-        // we can ignore any errors that come from this since they'll be
-        // re-emitted later
-        connection!.write(data, err => {
-          if (!err) {
-            this.latency_.add(Date.now() - startTime);
-          }
-          resolve();
-        });
-      });
-    });
+  private async _onflush(): Promise<void> {
```

Comment:
The mixing of `onFlush` and `onflush` is confusing. What up? :)

Reply:
Haha, whoops, I'll camelcase it.